x/twap/logic.go

```diff
@@ -144,15 +144,31 @@ func (k Keeper) updateRecords(ctx sdk.Context, poolId uint64) error {
 	}
 
 	for _, record := range records {
-		newRecord := k.updateRecord(ctx, record)
+		newRecord, err := k.updateRecord(ctx, record)
+		if err != nil {
+			return err
+		}
 		k.storeNewRecord(ctx, newRecord)
 	}
 	return nil
 }
 
 // updateRecord returns a new record with updated accumulators and block time
 // for the current block time.
-func (k Keeper) updateRecord(ctx sdk.Context, record types.TwapRecord) types.TwapRecord {
+func (k Keeper) updateRecord(ctx sdk.Context, record types.TwapRecord) (types.TwapRecord, error) {
+	// Returns error for last updated records in the same block.
+	// Exception: record is initialized when creating a pool,
+	// then the TwapAccumulator variables are zero.
+	if (record.Height >= ctx.BlockHeight() || !record.Time.Before(ctx.BlockTime())) &&
+		!record.P1ArithmeticTwapAccumulator.Equal(sdk.ZeroDec()) &&
```

Comment:
Can we add test cases to verify all these logic gates?
Right now only one test case has been added where we test if record.Height ctx.BlockHeight, but it'd be great if we could have additional tests for
1. same test but with record
2. same test as above but with P1ArithmeticTwapAccumulator ==0 and P0ArithmeticTwapAccumulator ==0
3. same test as above but with P1ArithmeticTwapAccumulator ==0 and P0ArithmeticTwapAccumulator !=0